Initial compliance-client script

akofink opened the pull request:
I'm really not very familiar with python packaging (like setup.py or even folder structure), so please voice any comments/concerns/improvements you have!

TODO:

- [x] This needs an update to the Readme with instructions on how to build/run/develop.
- [ ] Use config files and remove hardcoded values

Example running:

```
$ sudo UPLOAD_ARGS="--no-gpg" BYPASS_GPG=true x-rh-identity=eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6IjEyMTI3MjkiLCJpbnRlcm5hbCI6eyJvcmdfaWQiOiI1NDMyMSJ9LCJ1c2VyIjp7InVzZXJuYW1lIjoicmhuLWVuZ2luZWVyaW5nLWFrb2ZpbmsifX19 python /home/vagrant/compliance-client/compliance_client/__init__.py
Running scan... this may take a while
Uploading Insights data.
Successfully uploaded report for f1809dc8-2870-4d63-ab46-db7545b5fe56.
```

akofink commented:
@dLobatog well, this works now end to end. I'm not sure how to test it behind 3scale; you shouldn't need the x-rh-identity header env variable, just basic auth or cert-based auth like the normal insights-client uses. It's still very hacky, and I welcome any comments :)

quarckster commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,76 @@
+        return []
+
+def os_release():
+    grep = Popen("grep 7\.[0-9]* /etc/redhat-release", shell=True, stdout=PIPE, stderr=STDOUT).stdout.read()
```
Using shell=True can be a security hazard https://docs.python.org/2/library/subprocess.html#frequently-used-arguments. And instead of parsing `/etc/redhat-release` use `platform` module:
```
from platform import linux_distribution

linux_distribution()
```

quarckster commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,76 @@
+
+def find_scap_policy(profile_ref_id):
+    filenames = []
+    grep = Popen("grep {} /usr/share/xml/scap/ssg/content/*{}*.xml".format(profile_ref_id, os_release()), shell=True, stdout=PIPE, stderr=STDOUT).stdout.read()
```
```python
grep = Popen(["grep", profile_ref_id, "/usr/share/xml/scap/ssg/content/*{}*.xml".format(os_release())], stdout=PIPE, stderr=STDOUT).stdout.read()
```

quarckster commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,76 @@
+
+def run_scan(profile_ref_id, policy_xml):
+    print("Running scan... this may take a while")
+    Popen("oscap xccdf eval --profile {} --results {} {}".format(profile_ref_id, OSCAP_RESULTS_OUTPUT, policy_xml), shell=True, stdout=PIPE,
```
`shell=True`

quarckster commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,76 @@
+            stderr=STDOUT).stdout.read() # this will wait until the command finishes
+
+def upload_results():
+    return Popen('BYPASS_GPG=true insights-client --no-gpg --payload {} --content-type application/vnd.redhat.compliance.something+tgz'.format(OSCAP_RESULTS_OUTPUT),
```
```python
return Popen(['insights-client', '--no-gpg', '--payload', OSCAP_RESULTS_OUTPUT, '--content-type', 'application/vnd.redhat.compliance.something+tgz'], stdout=PIPE, stderr=STDOUT, env={'BYPASS_GPG': 'true'}).stdout.read()
```

quarckster commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,76 @@
+
+    fqdn = get_canonical_facts()['fqdn']
+    policies = get_policies(conn, fqdn)
+    if len(policies) == 0:
```
```python
if policies:
```

akofink commented:
Updated. Now you must pass `BYPASS_GPG=true` to compliance-client if you want it to be passed during the insights-client run. In addition, I've added the `UPLOAD_ARGS` env variable to pass any additional arguments to the insights client, such as `--no-gpg`, which is only used in non-standard or devel environments (see PR description for an example).

akofink commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,83 @@
+
+OSCAP_RESULTS_OUTPUT = '/tmp/oscap_results.xml'
+UPLOAD_ARGS = filter(None, [environ.get('UPLOAD_ARGS', None)])
+NONCOMPLIANT_STATUS = 2
```
oscap returns 1 if the scan fails. It returns 2 if the scan succeeds but the host is non-compliant.

quarckster commented on compliance_client/__init__.py:
```diff
@@ -0,0 +1,83 @@
+def main():
+    config = get_config()
+    conn = InsightsConnection(config)
+    conn.session.headers['x-rh-identity'] = environ['x-rh-identity'] if 'x-rh-identity' in environ else ''
```
```python
conn.session.headers['x-rh-identity'] = environ.get('x-rh-identity', '')
```

akofink commented:
Updated

akofink commented:
Closing in favor of https://github.com/RedHatInsights/insights-core/pull/1989
